spec/abstract/AARCH64/Arch_Structs_A.thy

```diff
@@ -305,6 +342,12 @@ definition asid_pool_level :: vm_level where
 definition max_pt_level :: vm_level where
   "max_pt_level = asid_pool_level - 1"
 
+definition level_type :: "vm_level \<Rightarrow> pt_type" where
+  "level_type level \<equiv> if level = max_pt_level then VSRootPT_T else NormalPT_T"
+
+declare [[coercion_enabled]]
+declare [[coercion level_type]]
```

Comment:
This is "Coercive subtyping" (page 297 in the Isar reference manual (isar-ref) for 2021-1, available in the trusty documentation tab in jEdit). It's a mechanism that upon type failure where `'a` is expected but ``b` found tries to find a coercion function `'a -> 'b` and inserts that as needed into the term. The theory behind it is a bit more complex, but this is pretty much what it does. It's similar to automatic type conversion in C where you can assign a `char` to and `int` and it will perform a cast.

In our case, if a term contains type `vmlevel` where type `pt_type` is expected, Isabelle will now automatically insert the function `level_type` to convert from one to the other. This is for input only, in output the type coercion function (here `level_type`) is always shown.

We haven't been using this feature so far, because things can become pretty confusing quickly when this is overused, but I think in this case it is beneficial. We could also use it for other things like `unat`, but let's see how this one goes first.

Reply:
Great, thanks! Unfortunately, the text in those railroad diagrams aren't searchable in the pdf of the reference manual.

Reply:
As always in Isabelle, the reference manual is easiest to use if you already know the answer :-)